pyartcd/pyartcd/pipelines/promote.py

```diff
@@ -375,6 +401,215 @@ def _reraise_if_not_permitted(self, err: VerificationError, code: str, permits:
+                shasum = hashlib.sha256(f.read()).hexdigest()
+            with open("sha256sum.txt", 'a') as f:  # write shasum to sha256sum.txt
+                f.write(f"{shasum}  opm-{platform}-{release_name}.tar.gz\n")
+        os.chdir(current_path)
+
+    async def publish_multi_client(self, working_dir, from_release_tag, release_name, client_type):
+        # Anything under this directory will be sync'd to the mirror
+        BASE_TO_MIRROR_DIR = f"{working_dir}/to_mirror/openshift-v4"
+        shutil.rmtree(f"{BASE_TO_MIRROR_DIR}/multi", ignore_errors=True)
+        RELEASE_MIRROR_DIR = f"{BASE_TO_MIRROR_DIR}/multi/clients/{client_type}/{release_name}"
+        current_path = os.getcwd()
+
+        for go_arch in util.go_arches:
+            if go_arch == "multi":
+                continue
+            # From the newly built release, extract the client tools into the workspace following the directory structure
+            # we expect to publish to mirror
+            CLIENT_MIRROR_DIR = f"{RELEASE_MIRROR_DIR}/{go_arch}"
+            os.makedirs(CLIENT_MIRROR_DIR)
+            # extract release clients tools
+            extract_release_client_tools(f"{constants.QUAY_RELEASE_REPO_URL}:{from_release_tag}", f"--to={CLIENT_MIRROR_DIR}", go_arch)
+            # create symlink for clients
+            self.create_symlink(path_to_dir=CLIENT_MIRROR_DIR, log_tree=True, log_shasum=True)
+
+        # Create a master sha256sum.txt including the sha256sum.txt files from all subarches
+        # This is the file we will sign -- trust is transitive to the subarches
+        os.chdir(RELEASE_MIRROR_DIR)
+        for dir in os.listdir():
+            if not os.path.isdir(dir):
+                continue
+            for root, dirs, files in os.walk(dir):
+                if "sha256sum.txt" not in files:
+                    continue
+                with open(f"{root}/sha256sum.txt", "rb") as f:
+                    shasum = hashlib.sha256(f.read()).hexdigest()
+                with open("sha256sum.txt", 'a') as f:  # write shasum to sha256sum.txt
+                    f.write(f"{shasum}  {dir}/sha256sum.txt\n")
+        os.chdir(current_path) # return to parent path
+        util.log_dir_tree(RELEASE_MIRROR_DIR)
+
+        # Publish the clients to our S3 bucket.
+        await exectools.cmd_assert_async(f"aws s3 sync --no-progress --exact-timestamps {BASE_TO_MIRROR_DIR}/multi s3://art-srv-enterprise/pub/openshift-v4/multi", stdout=sys.stderr)
+
+    def create_symlink(self, path_to_dir, log_tree, log_shasum):
+        # External consumers want a link they can rely on.. e.g. .../latest/openshift-client-linux.tgz .
+        # So whatever we extract, remove the version specific info and make a symlink with that name.
+        current_path = os.getcwd() # /mnt/workspace/jenkins/working/build_promote-assembly
+        os.chdir(path_to_dir) # path_to_dir is relative path artcd_working/to_mirror/openshift-v4/aarch64/clients/ocp/4.13.0-rc.6
```

Comment:
Avoid os.chdir if possible. This could lead to race condition if we run those functions with multithread or coroutine in the future.

Reply:
same as above